File: FloatingShortcutsPhone/src/main/java/net/geekstools/floatshort/PRO/Widget/NavAdapter/InstalledWidgetsAdapter.java
package net.geekstools.floatshort.PRO.Widget.NavAdapter;

import android.app.Activity;
import android.appwidget.AppWidgetHost;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProviderInfo;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import net.geekstools.floatshort.PRO.R;
import net.geekstools.floatshort.PRO.Util.Functions.FunctionsClass;
import net.geekstools.floatshort.PRO.Util.Functions.FunctionsClassDebug;
import net.geekstools.floatshort.PRO.Util.Functions.PublicVariable;
import net.geekstools.floatshort.PRO.Util.NavAdapter.NavDrawerItem;

import java.util.ArrayList;

public class InstalledWidgetsAdapter extends RecyclerView.Adapter<InstalledWidgetsAdapter.ViewHolder> {

    Activity activity;
    Context context;

    FunctionsClass functionsClass;

    ArrayList<NavDrawerItem> navDrawerItems;

    View view;
    ViewHolder viewHolder;

    public static final int WIDGET_CONFIGURATION_REQUEST = 666;
    public static final int SYSTEM_WIDGET_PICKER = 333;
    public static final int SYSTEM_WIDGET_PICKER_CONFIGURATION = 111;

    public static String pickedWidgetPackageName, pickedWidgetClassNameProvider, pickedWidgetConfigClassName = null;
    public static AppWidgetProviderInfo pickedAppWidgetProviderInfo;
    public static int pickedWidgetId;
    public static String pickedWidgetLabel;

    AppWidgetHost appWidgetHost;

    public InstalledWidgetsAdapter(Activity activity, Context context, ArrayList<NavDrawerItem> navDrawerItems, AppWidgetHost appWidgetHost) {
        this.context = context;
        this.activity = activity;

        this.navDrawerItems = navDrawerItems;

        this.appWidgetHost = appWidgetHost;

        functionsClass = new FunctionsClass(context, activity);
    }

    @Override
    public InstalledWidgetsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        view = LayoutInflater.from(context).inflate(R.layout.widget_installed_items, parent, false);
        viewHolder = new ViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolderBinder, final int position) {

        viewHolder.widgetPreview.setImageDrawable(navDrawerItems.get(position).getWidgetPreview());
        viewHolder.widgetLabel.setText(navDrawerItems.get(position).getWidgetLabel());
        viewHolder.widgetLabel.setTextColor(PublicVariable.themeLightDark ? context.getColor(R.color.dark) : context.getColor(R.color.light));

        viewHolder.widgetitem.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

            }
        });
        viewHolder.widgetitem.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                functionsClass.doVibrate(77);

                pickedWidgetPackageName = navDrawerItems.get(position).getPackageName();
                pickedWidgetClassNameProvider = navDrawerItems.get(position).getClassNameProviderWidget();
                pickedWidgetConfigClassName = navDrawerItems.get(position).getConfigClassNameWidget();
                pickedAppWidgetProviderInfo = navDrawerItems.get(position).getAppWidgetProviderInfo();
                pickedWidgetId = appWidgetHost.allocateAppWidgetId();
                pickedWidgetLabel = navDrawerItems.get(position).getWidgetLabel();

                ComponentName provider = ComponentName.createRelative(pickedWidgetPackageName, pickedWidgetClassNameProvider);
                FunctionsClassDebug.Companion.PrintDebug("*** Provider Widget = " + provider);

                Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_BIND);
                intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, pickedWidgetId);
                intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, provider);
                activity.startActivityForResult(intent, WIDGET_CONFIGURATION_REQUEST);

                if (pickedAppWidgetProviderInfo.configure != null) {
                    ComponentName configure = ComponentName.createRelative(pickedWidgetPackageName, pickedWidgetConfigClassName);
                    FunctionsClassDebug.Companion.PrintDebug("*** Configure Widget = " + configure);

                    Intent intentWidgetConfiguration = new Intent();
                    intentWidgetConfiguration.setAction(AppWidgetManager.ACTION_APPWIDGET_CONFIGURE);
                    intentWidgetConfiguration.setComponent(/*pickedAppWidgetProviderInfo.*/configure);
                    intentWidgetConfiguration.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, pickedWidgetId);
                    intentWidgetConfiguration.putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, /*pickedAppWidgetProviderInfo.*/provider);
                    intentWidgetConfiguration.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_NO_HISTORY | Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
                    activity.startActivityForResult(intentWidgetConfiguration, WIDGET_CONFIGURATION_REQUEST);
                }

                return false;
            }
        });
    }

    @Override
    public int getItemCount() {
        return navDrawerItems.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        RelativeLayout widgetitem;
        ImageView widgetPreview;
        TextView widgetLabel;

        public ViewHolder(View view) {
            super(view);
            widgetitem = (RelativeLayout) view.findViewById(R.id.widgetItem);
            widgetPreview = (ImageView) view.findViewById(R.id.widgetPreview);
            widgetLabel = (TextView) view.findViewById(R.id.widgetLabel);
        }
    }
}

File: FloatingMultitasking/src/main/java/net/geekstools/floatshort/PRO/Widgets/WidgetsAdapter/InstalledWidgetsAdapter.kt
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel
 * Last modified 8/29/20 3:58 AM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */
package net.geekstools.floatshort.PRO.Widgets.WidgetsAdapter

import android.appwidget.AppWidgetHost
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProviderInfo
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import net.geekstools.floatshort.PRO.R
import net.geekstools.floatshort.PRO.Utils.AdapterItemsData.AdapterItems
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClassLegacy
import net.geekstools.floatshort.PRO.Utils.Functions.PublicVariable
import net.geekstools.floatshort.PRO.Widgets.WidgetConfigurations

class InstalledWidgetsAdapter(private val widgetConfigurationsActivity: WidgetConfigurations, private val context: Context,
                              private val adapterItems: ArrayList<AdapterItems>, private val appWidgetHost: AppWidgetHost) : RecyclerView.Adapter<InstalledWidgetsAdapter.ViewHolder>() {

    private val functionsClassLegacy: FunctionsClassLegacy = FunctionsClassLegacy(context)

    companion object {
        const val WIDGET_CONFIGURATION_REQUEST = 666
        const val SYSTEM_WIDGET_PICKER = 333
        const val SYSTEM_WIDGET_PICKER_CONFIGURATION = 111

        var pickedWidgetPackageName: String? = null
        var pickedWidgetClassNameProvider: String? = null
        var pickedWidgetConfigClassName: String? = null
        lateinit var pickedAppWidgetProviderInfo: AppWidgetProviderInfo

        var pickedWidgetId = 0
        lateinit var pickedWidgetLabel: String
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.widget_installed_items, parent, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {

        viewHolder.widgetPreview.setImageDrawable(adapterItems[position].widgetPreview)
        viewHolder.widgetLabel.text = adapterItems[position].widgetLabel
        viewHolder.widgetLabel.setTextColor(if (PublicVariable.themeLightDark) context.getColor(R.color.dark) else context.getColor(R.color.light))

        viewHolder.widgetitem.setOnClickListener {

        }

        viewHolder.widgetitem.setOnLongClickListener {
            functionsClassLegacy.doVibrate(77)

            InstalledWidgetsAdapter.pickedWidgetPackageName = adapterItems[position].packageName
            InstalledWidgetsAdapter.pickedWidgetClassNameProvider = adapterItems[position].classNameProviderWidget
            InstalledWidgetsAdapter.pickedWidgetConfigClassName = adapterItems[position].configClassNameWidget
            InstalledWidgetsAdapter.pickedAppWidgetProviderInfo = adapterItems[position].appWidgetProviderInfo

            InstalledWidgetsAdapter.pickedWidgetId = appWidgetHost.allocateAppWidgetId()
            InstalledWidgetsAdapter.pickedWidgetLabel = adapterItems[position].widgetLabel


            if (InstalledWidgetsAdapter.pickedWidgetPackageName != null &&
                    InstalledWidgetsAdapter.pickedWidgetClassNameProvider != null) {

                val widgetProvider = ComponentName.createRelative(InstalledWidgetsAdapter.pickedWidgetPackageName!!, InstalledWidgetsAdapter.pickedWidgetClassNameProvider!!)

                if (InstalledWidgetsAdapter.pickedAppWidgetProviderInfo.configure != null
                        && InstalledWidgetsAdapter.pickedWidgetConfigClassName != null) {


                    try {
                        val widgetConfigurationActivity = ComponentName.createRelative(InstalledWidgetsAdapter.pickedWidgetPackageName!!, InstalledWidgetsAdapter.pickedWidgetConfigClassName!!)

                        if (context.packageManager.getActivityInfo(widgetConfigurationActivity, PackageManager.GET_META_DATA).exported) {

                            val intentWidgetBind = Intent(AppWidgetManager.ACTION_APPWIDGET_BIND).apply {
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, pickedWidgetId)
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, widgetProvider)
                            }
                            widgetConfigurationsActivity.startActivityForResult(intentWidgetBind, InstalledWidgetsAdapter.WIDGET_CONFIGURATION_REQUEST)

                            val intentWidgetConfiguration = Intent().apply {
                                action = AppWidgetManager.ACTION_APPWIDGET_CONFIGURE
                                component = widgetConfigurationActivity
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, pickedWidgetId)
                                putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER,  /*pickedAppWidgetProviderInfo.*/widgetProvider)
                                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_NO_HISTORY or Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS)
                            }
                            widgetConfigurationsActivity.startActivityForResult(intentWidgetConfiguration, InstalledWidgetsAdapter.WIDGET_CONFIGURATION_REQUEST)

                        } else {

                        }
                    } catch (e: PackageManager.NameNotFoundException) {
                        e.printStackTrace()
                    }

                } else {

                    val intent = Intent(AppWidgetManager.ACTION_APPWIDGET_BIND)
                    intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, pickedWidgetId)
                    intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_PROVIDER, widgetProvider)
                    widgetConfigurationsActivity.startActivityForResult(intent, InstalledWidgetsAdapter.WIDGET_CONFIGURATION_REQUEST)
                }
            }

            false
        }
    }

    override fun getItemCount(): Int {

        return adapterItems.size
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var widgetitem: RelativeLayout = view.findViewById<RelativeLayout>(R.id.widgetItem) as RelativeLayout
        var widgetPreview: ImageView = view.findViewById<ImageView>(R.id.widgetPreview) as ImageView
        var widgetLabel: TextView = view.findViewById<TextView>(R.id.widgetLabel) as TextView
    }
}